### PR TITLE
Fix: Use the correct location for `comma-dangle` errors (fixes #7291)

### LINKED ISSUE #7291: comma-dangle reports incorrect position in array with ternary item
**Tell us about your environment**

* **ESLint Version:** `3.7.0`
* **Node Version:** `4.2.2`
* **npm Version:** `3.10.7`

**What parser (default, Babel-ESLint, etc.) are you using?**

`babel-eslint@7.0.0`

**Please show your full configuration:**

```
module.exports = {
  "extends": "eslint-config-airbnb",
  "env": {
    "browser": true,
    "node": true,
    "mocha": true
  },
  "parser": "babel-eslint",
  "rules": {
    "no-unused-vars": "warn",
    "no-alert": "warn",
    "no-console": "warn",
    "no-use-before-define": "warn",
    "no-var": "warn",
    "block-scoped-var": "warn",
    "guard-for-in": "off",
    "no-shadow": "off",
    "no-param-reassign": "off",

    "indent": [ "warn", 2, { "SwitchCase": 1 } ],
    "padded-blocks": "off",
    "spaced-comment": "warn",
    "comma-dangle": [ "warn", "always-multiline" ],
    "space-in-parens": "off",
    "curly": [ "off", "all" ],
    "array-bracket-spacing": [ "warn", "always" ],
    "object-curly-spacing": [ "warn", "always" ],
    "computed-property-spacing": "off",
    "brace-style": [ "warn", "stroustrup", { "allowSingleLine": true } ],
    "no-trailing-spaces": [ "warn", { "skipBlankLines": true } ],
    "linebreak-style": [ "warn", "unix" ],
    "no-multiple-empty-lines": [ "warn", { "max": 2, "maxEOF": 1 } ],
    "eol-last": [ "error", "unix" ],
    "id-length": [ "warn", { "min": 2, "exceptions": [ "_", "$", "i", "j", "k", "x", "y", "e" ] } ],
    "camelcase": "warn",
    "func-names": "warn",
    "keyword-spacing": "warn",
    "space-before-blocks": "warn",
    "space-before-function-paren": [ "warn", { "anonymous": "always", "named": "never" } ],
    "quotes": [ "warn", "single", "avoid-escape" ],
    "no-multi-spaces": [ "warn", { "exceptions": { "VariableDeclarator": true, "ImportDeclaration": true } } ],
    "dot-notation": "off",
    "prefer-template": "off",
    "prefer-arrow-callback": "warn",
    "max-len": [ "warn", 250, 4, { "ignoreComments": true } ],
    "arrow-parens": [ "warn", "always" ],
    "arrow-body-style": "off",
    "object-shorthand": "off",
    "no-case-declarations": "warn",
    "no-nested-ternary": "off",
    "global-require": "off",
    "no-underscore-dangle": "off",
    "no-useless-concat": "off",
    "no-mixed-operators": "off",
    "no-bitwise": "off",

    "react/prop-types": "warn",
    "react/forbid-prop-types": "off",
    "react/no-unused-prop-types": "off",
    "react/react-in-jsx-scope": "error",
    "react/jsx-filename-extension": [ "warn", { "extensions": [ ".js", ".jsx" ] } ],
    "react/jsx-uses-react": "error",
    "react/jsx-uses-vars": "error",
    "react/jsx-quotes": "off",
    "react/jsx-closing-bracket-location": "warn",
    "react/jsx-space-before-closing": "warn",
    "react/jsx-curly-spacing": "off",
    "react/jsx-indent": "off",
    "react/self-closing-comp": "warn",
    "react/no-multi-comp": "off",
    "react/sort-comp": "off",
    "react/prefer-stateless-function": "warn",
    "react/no-children-prop": "off",
    "jsx-quotes": "error",
    "jsx-a11y/img-uses-alt": "off",

    "import/default": "off",
    "import/no-duplicates": "off",
    "import/named": "off",
    "import/namespace": "off",
    "import/newline-after-import": "off",
    "import/imports-first": "off",
    "import/no-unresolved": "error",
    "import/no-named-as-default": "error",
    "import/extensions": [ "warn", "always", { "": "never", "js": "never" } ],
  },
  "plugins": [
    "react",
    "import",
  ],
  "settings": {
    "import/ignore": [
      "node_modules",
      "\\.(scss|less|css)$",
    ],
    "import/resolver": {
      "node": {
        "moduleDirectory": [
          "./src",
          "./api",
          "./common",
        ],
      },
      "webpack": {
        "config": "./webpack/configForEslintImportResolver.js",
      },
    },
  },
  "globals": {
    "__SERVER_ENV__": true,
    "__DEVELOPMENT__": true,
    "__SERVER__": true,
    "__SERVER_DEVTOOLS__": true,
    "__SERVER_DISABLE_SERVERSIDE_RENDERING__": true,
    "__CLIENT__": true,
    "__CLIENT_DEVTOOLS__": true,
    "__LOCALHOST__": true,
    "__FAST_DEVELOPMENT__": true,
  },
};
```

**What did you do? Please include the actual source code causing the issue.**

Reduced test case:

```js
let xxx; let yyy;
const zzz = [
  (xxx
    ? yyy
    : ''   // <<< reports that the comma is missing here
  )   // <<< should report the missing comma here
];
zzz.toString();  // to prevent unused variable warning in the reduced test case

```

**What did you expect to happen?**

Correct positioning of the warning (after the closing paren where the comma-dangle is expected).

**What actually happened? Please include the actual, raw output from ESLint.**

```
  5:13  warning  Missing trailing comma  comma-dangle
```

### PR DESCRIPTION
**Pull request by not-an-aardvark**
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[x] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

See #7291

**Please check each item to ensure your pull request is ready:**

- [x] I've read the [pull request guide](http://eslint.org/docs/developer-guide/contributing/pull-requests)
- [x] I've included tests for my change
- (n/a) I've updated documentation for my change (if appropriate)

**What changes did you make? (Give an overview)**

Previously, `comma-dangle` would report the location of the last node in an array/object if a trailing comma was missing. As a result, the error would sometimes be reported on the wrong line if the node was parenthesized.

This PR fixes the error location; now it always reports the location of the token before the closing `]` or `}`.

**Is there anything you'd like reviewers to focus on?**

Nothing in particular.

### CODE REVIEW

**Comment by platinumazure** [comment]
What will this do to auto-fix, if anything?

**Comment by not-an-aardvark** [comment]
I don't think it will have any effect on auto-fixing, since the same fixer function is returned either way.

File: tests/lib/rules/comma-dangle.js [hunk 1]
```diff
@@ -392,8 +392,8 @@ ruleTester.run("comma-dangle", rule, {
                 {
                     message: "Missing trailing comma.",
                     type: "Property",
-                    line: 2,
-                    column: 11
+                    line: 3,
+                    column: 1
```
Comment (vitorbal) [inline tests/lib/rules/comma-dangle.js]: I don't know if I agree with changing the reported location for these test cases. I think it makes more sense from a user perspective that the reported location is right after the property instead of at the closing brace. 
An example is the test case marked by this comment:
Before:
```bash
  1 | var foo = {
> 2 | bar: 'baz'
    |           ^
  3 | }
```
After:
```bash
  1 | var foo = {
  2 | bar: 'baz'
> 3 | }
    | ^
```
I think the original reported location in this case makes more logical sense for the user, since that's where he will have to insert the trailing comma.

What if you would report on `penultimateToken.loc.end` instead of on `trailingToken.loc.start`? I think that would fix the edge case of this bug report and preserve the location reported on all the other tests cases.

**Comment by vitorbal** [comment]
Could you add some regression test cases for the examples that would trigger the bug? Here is a couple that come to mind:
```javascript
        {
            code: "var zzz = [\n(xxx\n? yyy\n: ''\n)\n];",
            output: "var zzz = [\n(xxx\n? yyy\n: ''\n),\n];",
            options: [ "always" ],
            errors: [
                {
                    message: "Missing trailing comma.",
                    type: "ConditionalExpression",
                    line: 5,
                    column: 2
                }
            ]
        },
        {
            code: "var zzz = [\n(\n'zzz'\n)\n];",
            output: "var zzz = [\n(\n'zzz'\n),\n];",
            options: [ "always" ],
            errors: [
                {
                    message: "Missing trailing comma.",
                    type: "Literal",
                    line: 4,
                    column: 2
                }
            ]
        },
```

File: tests/lib/rules/comma-dangle.js (hunk 1 above)
Comment (not-an-aardvark) [inline tests/lib/rules/comma-dangle.js]: > I think the original reported location in this case makes more logical sense for the user, since that's where he will have to insert the trailing comma.

As a nitpick, this is only true if the user is using `comma-style: last`. If the user formatted their commas like this, then reporting the closing bracket would make more sense:

```js
var foo = {
  bar: 'baz'
  , qux: 'foobar'
};
^
```

However, in general I would say you're probably right, since I think `comma-style: last` is much more common.

**Comment by not-an-aardvark** [comment]
I updated the PR to avoid modifying the location for existing test cases, and added some new ones with `penultimateToken.loc.end` reported.

File: tests/lib/rules/comma-dangle.js (hunk 1 above)
Comment (vitorbal) [inline tests/lib/rules/comma-dangle.js]: Ah, good point about `comma-style`. Agree with you, though, that `comma-style: last` is the more commonly used config.

**Comment by kaicataldo** [comment]
Sorry, didn't mean to delete your branch - habit from merging my own PRs...

**Comment by sompylasar** [comment]
Yay thanks!
